### PR TITLE
Integration of Fidelity Kernel

### PR DESCRIPTION
**Pull request by CassNot**
# Summary

- Introduces a more robust `FeatureMap` with flexible encoding and better dtype/device handling.
- Adds `KernelCircuitBuilder` for fluent construction of feature maps and fidelity kernels from photonic backends.
- Refactors `FidelityKernel` to support batched computation, optional sampling, and PSD projection for ML compatibility.

## Context

- Follow-up to draft PR #7 on the upstream repo.
- Supersedes #7 and incorporates its goals with additional fixes, tests, and docs.

### dtype/device
- Centralized `dtype_to_torch` mapping for numpy/torch types and strings.
- Consistent device/dtype normalization across all entry points.

### FeatureMap
- New `input_size` argument and validation helpers (`is_datapoint`).
- Flexible feature encoding with 

1. Optional encoder hook (auto-discovered from `Ansatz`)
2. Deterministic subset-sum expansion to match circuit parameter length (for the Series encoding, to avoid an Index Error)
3. Safe handling of scalars/arrays; truncation/padding logic.

- Trainable parameters are registered and passed to `CircuitConverter`.
- Factory methods: 

1. `from_photonic_backend`: builds from `PhotonicBackend` via `AnsatzFactory` and auto-wires encoder
2. `simple`: convenience factory for minimal config (to stay in the spirit of the "no quantum knowledge required"

- `KernelCircuitBuilder` (new)

1. Fluent builder for typical kernel setups (circuit type, modes, photons, state pattern, reservoir mode, dtype/device, bandwidth tuning).
2. `build_feature_map` and `build_fidelity_kernel` helpers with sensible defaults.

- `FidelityKernel`

1. Validates `input_state` vs circuit.
2. Batched unitary evaluation and SLOS calls: Vectorized U(x1) U(x2)† construction and Upper-triangle evaluation for symmetric training matrices.
3. Sampling support via `AutoDiffProcess` (multinomial/binomial/gaussian).
4. PSD projection option for stable downstream ML (SVM/SVC).
5. Convenience factories: `from_photonic_backend` and `simple`.
6. Safer equality checks and shape handling; scalar datapoint path.

- Requirements:

1. Added `scikit-learn` as a requirements when installing `merlinquantum`

#### A small and simple example !

```
import torch
from sklearn.datasets import load_iris
from sklearn.svm import SVC
from merlin.algorithms.kernels import FidelityKernel

iris = load_iris()
X, y = iris.data, iris.target
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

kernel = FidelityKernel.simple(
    input_size=4,  # number of features
    n_modes=6,  # photonic modes
    n_photons=2,  # photons (<= n_modes)
    circuit_type="series",
    reservoir_mode=False,
    force_psd=True,  # project to PSD if needed
)
# Evaluate
X_train_t = torch.tensor(X_train, dtype=torch.float32)
K_train = kernel(X_train_t).detach().numpy()
K_train = kernel(X_train_t).detach().numpy()
K_test = kernel(X_test_t, X_train_t).detach().numpy()
clf = SVC(kernel="precomputed", random_state=42)
clf.fit(K_train, y_train)
test_accuracy =  clf.score(K_test, y_test)
```

## Why

- Simplify and standardize how classical features are embedded into photonic circuits.
- Simplify kernel creation using `PhotonicBackend`/`Ansatz`.
- Improve throughput for kernel matrix builds via batching and reduced redundant compute.
- Ensure generated kernels are numerically sound and ML-ready.

## Performance/UX notes

- Batched computation reduces repeated SLOS graph invocations.
- Upper-triangular computation lowers cost for training Gram matrices.
- Device/dtype normalization prevents subtle CPU/GPU/dtype mismatches.

## Backward compatibility

- FeatureMap now requires input_size and only supports a single input parameter name (string or single-item list). If multiple parameter names were used previously, migrate to a single logical input parameter or provide a custom encoder.
- New builder is additive; FidelityKernel API remains compatible with additional conveniences.

## Testing/validation

- Manual checks on:

1. Scalar and batched inputs.
3. Sampling on/off paths.
4. PSD projection toggled.
5. no_bunching validation with different input states.
6. Test on a real-world scenario (IRIS dataset) with all builders

## Open questions for reviewers

- Should subset-sum expansion be optional/configurable?
- Do we want stricter behavior when encoded length mismatches (raise vs fallback)?

## Documentation

- Docstrings updated with usage examples.
- Short tutorial (notebook) showing `KernelCircuitBuilder` → `FidelityKernel` → `scikit-learn `SVC workflow

## Closes/Related

Related to improving kernel construction ergonomics and performance from photonic backends.

### CODE REVIEW

File: merlin/algorithms/kernels.py [hunk 1]
```diff
@@ -0,0 +1,878 @@
...
+        all_probs = self._slos_graph.compute(all_circuits, self.input_state)[1]
+
+        if self.shots > 0:
+            # Convert complex amplitudes to real probabilities for multinomial sampling
+            real_probs = torch.abs(all_probs)
+            all_probs = self._autodiff_process.sampling_noise.pcvl_sampler(
+                real_probs, self.shots, self.sampling_method
+            )
+
+        transition_probs = torch.abs(all_probs[:, self._input_state_index])
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/algorithms/kernels.py]: This is correct. I fixed this line by adding the square command. The probability = |complex amplitude|^2 !

**Comment by jsenellart** [comment]
as discussed, it would be good to use "Ansatz" as description of the circuit/backend for the kernel -  this will in particular enable running the kernel on a QPU - and will mutualize all the circuit definition with QuantumLayer. Contrary to QuantumLayer, since it is its first introduction, we should not keep the alternative definition.
